heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts

```diff
@@ -0,0 +1,246 @@
+    }
+    return await this._apiExtractorConfigurationFilePathPromise;
+  }
+
+  private async _getApiExtractorConfigurationAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    ignoreMissingEntryPoint?: boolean
+  ): Promise<TApiExtractor.ExtractorConfig | undefined> {
+    const configObjectFullPath: string | undefined = await this._getApiExtractorConfigurationFilePathAsync(
+      heftConfiguration
+    );
+    if (!configObjectFullPath) {
+      return undefined;
+    }
+
+    const apiExtractor: typeof TApiExtractor = (await this._getApiExtractorAsync(
+      taskSession,
+      heftConfiguration
+    ))!;
+    const configObject: TApiExtractor.IConfigFile =
+      apiExtractor.ExtractorConfig.loadFile(configObjectFullPath);
+
+    return apiExtractor.ExtractorConfig.prepare({
+      configObject,
+      configObjectFullPath,
+      ignoreMissingEntryPoint,
+      packageJsonFullPath: path.join(heftConfiguration.buildFolder, 'package.json'),
+      projectFolderLookupToken: heftConfiguration.buildFolder
+    });
+  }
+
+  private async _getApiExtractorAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration
+  ): Promise<typeof TApiExtractor | undefined> {
+    if (!this._apiExtractorPromise) {
+      this._apiExtractorPromise = this._getApiExtractorInnerAsync(taskSession, heftConfiguration);
+    }
+    return await this._apiExtractorPromise;
+  }
+
+  private async _getApiExtractorInnerAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration
+  ): Promise<typeof TApiExtractor | undefined> {
+    // API Extractor provides an ExtractorConfig.tryLoadForFolder() API that will probe for api-extractor.json
+    // including support for rig.json.  However, Heft does not load the @microsoft/api-extractor package at all
+    // unless it sees a config/api-extractor.json file.  Thus we need to do our own lookup here.
+    const apiExtractorConfigurationFilePath: string | undefined =
+      await this._getApiExtractorConfigurationFilePathAsync(heftConfiguration);
+    if (!apiExtractorConfigurationFilePath) {
+      return undefined;
+    }
+
+    const apiExtractorPackagePath: string = await heftConfiguration.rigToolResolver.resolvePackageAsync(
+      '@microsoft/api-extractor',
+      taskSession.logger.terminal
+    );
+    return await import(apiExtractorPackagePath);
+  }
+
+  private async _updateCleanOptionsAsync(
+    cleanOptions: IHeftTaskCleanHookOptions,
+    apiExtractorConfiguration: TApiExtractor.ExtractorConfig
+  ): Promise<void> {
+    const extractorGeneratedFilePaths: string[] = [];
+    if (apiExtractorConfiguration.apiReportEnabled) {
+      // Keep apiExtractorConfiguration.reportFilePath as-is, since API-Extractor uses the existing
+      // content to write a warning if the output has changed.
+      extractorGeneratedFilePaths.push(apiExtractorConfiguration.reportTempFilePath);
```

Comment:
This usually goes under the `temp` folder, which gets cleaned up anyway. Does the clean functionality ignore files that are under folders that are already getting deleted?

Reply:
No, the clean functionality doesn't ignore those files, but it does swallow any exceptions if the file cannot be found. Since we already operate under the guise of "run the operation and swallow not found errors" for other things, I figured this was fair enough here.

Note: We do de-dupe input paths, so we will only ever attempt to delete a path once.